splicemachine/spliceengine
# DB-11858 Query rewrite for UNION queries

## Short Description
Move tables with unreferenced columns from the FROM clause into EXISTS WHERE clause conditions.

## Long Description
This feature does a query rewrite of SetOp queries of the form:
> select t1.* from t1, t2 union
> select t3.* from t3;

to:

> select t1.* from t1 where exists (select 1 from t2) union
> select t3.* from t3;

This eliminates a potentially costly join if t2 is large.  This works because duplicates are removed from the final result set and if no new values are added from the table due to no referenced columns, it does not impact the values in the final set.

For SetOp queries of the form:
> select 1,2,3 from t1 union
> select t2.a, t2.b, t2,c from t2;

They are rewritten as:

> select TOP 1 1,2,3 from t1 union
> select t2.a, t2.b, t2,c from t2;

This reduces the rows read from t1.
The presence of aggregate functions, window functions or nondeterministic functions in a SetOp branch disqualifies that branch from this rewrite.  This rewrite does not apply to the 'ALL' SetOps: UNION ALL, INTERSECT ALL, EXCEPT ALL (the last 2 are not yet implemented in Splice).

## How to test

create table t1 (a int, b int);
create table t2 (a int, b int);
// The explain of the following should contain 'Limit' and 'Subquery', but no Join operations.
explain
select * from (select a.a from t2 a, t1 b union
               select 1 from t2);

Plan
//----
Cursor(n=16,rows=1024,updateMode=READ_ONLY (1),engine=OLTP (default))
  ->  ScrollInsensitive(n=16,totalCost=27.048,outputRows=1024,outputHeapSize=1.894 KB,partitions=1,parallelTasks=1)
    ->  Distinct(n=13,totalCost=5.315,outputRows=1024,outputHeapSize=1.894 KB,partitions=1,parallelTasks=1)
      ->  Union(n=11,totalCost=5.315,outputRows=1946,outputHeapSize=3.6 KB,partitions=2,parallelTasks=1)
        ->  Limit(n=10,totalCost=0.005,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1,fetchFirst=1)
          ->  ProjectRestrict(n=9,totalCost=0.005,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
            ->  TableScan[T2(1712)](n=7,totalCost=0.005,scannedRows=1,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
        ->  ProjectRestrict(n=5,totalCost=5.315,outputRows=922,outputHeapSize=3.6 KB,partitions=1,parallelTasks=1,preds=[is not null(subq=4)])
          ->  Subquery(n=4,totalCost=0.014,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1,correlated=false,expression=true,invariant=true)
            ->  Limit(n=4,totalCost=0.003,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1,fetchFirst=1)
              ->  ProjectRestrict(n=3,totalCost=0.003,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
                ->  TableScan[T1(1696)](n=1,totalCost=0.003,scannedRows=1,outputRows=1,outputHeapSize=0 B,partitions=1,parallelTasks=1)
          ->  TableScan[T2(1712)](n=0,totalCost=5.106,scannedRows=1024,outputRows=1024,outputHeapSize=3.6 KB,partitions=1,parallelTasks=1)]
